llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -462,21 +530,35 @@ void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
 /// metadata and if so collects aspects from this metadata
 void processFunction(Function &F, FunctionToAspectsMapTy &FunctionToUsedAspects,
                      FunctionToAspectsMapTy &FunctionToDeclaredAspects,
-                     TypeToAspectsMapTy &TypesWithAspects, CallGraphTy &CG) {
+                     TypeToAspectsMapTy &TypesWithAspects, CallGraphTy &CG,
+                     const AspectValueToNameMapTy &AspectValues,
+                     bool FP64ConvEmu) {
+  auto FP64AspectIt = AspectValues.find("fp64");
+  assert(FP64AspectIt != AspectValues.end() &&
+         "fp64 aspect was not found in the aspect values.");
+  auto FP64Aspect = FP64AspectIt->second;
   const AspectsSetTy RetTyAspects =
       getAspectsFromType(F.getReturnType(), TypesWithAspects);
-  FunctionToUsedAspects[&F].insert(RetTyAspects.begin(), RetTyAspects.end());
+  for (const auto &Aspect : RetTyAspects)
+    if (!hasDoubleType(F.getReturnType()) || !FP64ConvEmu ||
+        (Aspect != FP64Aspect))
+      FunctionToUsedAspects[&F].insert(Aspect);
   for (Argument &Arg : F.args()) {
     const AspectsSetTy ArgAspects =
         getAspectsFromType(Arg.getType(), TypesWithAspects);
-    FunctionToUsedAspects[&F].insert(ArgAspects.begin(), ArgAspects.end());
+    for (const auto &Aspect : ArgAspects)
+      if (!hasDoubleType(Arg.getType()) || !FP64ConvEmu ||
```

Comment:
The first check in the 'if' condition has different inputs in the three different places. So, it might a bit counter-productive to try to combine them.

Thanks